sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -967,6 +969,23 @@ class DataFrame private[sql](
   }
 
   /**
+   * Randomly splits this DataFrame with the provided weights.
+   *
+   * @param weights weights for splits, will be normalized if they don't sum to 1
+   * @param seed random seed
+   *
+   * @return split DataFrames in an array
+   */
+  def randomSplit(weights: Array[Double], seed: Long = Utils.random.nextLong): Array[DataFrame] = {
+    val sum = weights.sum
+    val normalizedCumWeights = weights.map(_ / sum).scanLeft(0.0d)(_ + _)
+    normalizedCumWeights.sliding(2).map { x =>
+      this.sqlContext.createDataFrame(new PartitionwiseSampledRDD[Row, Row](
+        rdd, new BernoulliCellSampler[Row](x(0), x(1)), true, seed), schema)
```

Comment:
this actually breaks the plan -- can we create a logical operator (or generalizes the existing Sample operator) so the returned DataFrame correctly preserves the logical plan?